source/mongodb-handshake/handshake.rst

```diff
@@ -224,6 +224,49 @@ Example::
         - "Oracle JVM EE 9.1.1"
 
 
+Supporting Wrapping Libraries
+=============================
+
+Drivers MUST allow libraries which wrap the driver to append to the client
+metadata generated by the driver. The following class definition defines the
+options which MUST be supported:
+
+.. code:: typescript
+
+  class DriverInfoOptions {
+    /**
+     * The name of the library wrapping the driver.
+     */
+    name: Optional<String>;
```

Comment:
What if there are multiple levels of wrapping?  For example, in Java the Scala driver wraps the reactive streams driver, which wraps the core driver.  The Java driver handles that by making each of these fields by a list rather than a single value.  We can't realistically remove that from the Java driver, and it may be useful for others as well, e.g. maybe PyModm will wrap Motor some day, idk.

Reply:
I'd lean towards "defy augury" here, but also because I'm not really sure I could imagine what that would look like. Maybe it makes sense if we owned all the component pieces, but out in the community you would have to plumb down support for this all the way through the stack (Wrapper A wraps Driver and provides its own API for things that wrap _it_, etc). Who ultimately appends to the list in these options? I would imagine a wrapping driver is not interested in exposing access to its internal `MongoClient`, but I might just be making a bad assumption here.

Reply:
Oh, augury...  Java driver may be special here in terms of the number of layers that are all controlled by us.  So long as you word it such that the Java driver is not violating the spec doing what it's currently doing, I'm ok with it.

Reply:
> We can't realistically remove that from the Java driver, and it may be useful for others as well, e.g. maybe PyModm will wrap Motor some day, idk.

libmongoc exposes an API that allows whatever sits atop it to append a driver name/version. In turn, I would expect the PHP driver to provide an API to whatever sits atop it to append to the driver name/version. The PHP driver would just ensure that its own name/version ends up appearing before the next higher-level thing, and utilize string concatenation rather than tracking lists of things.

I imagine that's what Matt might have been hinting at with "plumb down support for this all the way through the stack", but I agree however Java currently handles this is fine provided its API accomplishes the same outcome.

Given that some drivers already have existing APIs to do this (e.g. PyMongo, libmongoc, Ruby), perhaps we can allow utmost flexibility and just propose DriverInfoOptions as one possible API. The MUST and MUST NOTs can then just refer to rules for assembling the driver name/version and platform strings.

Reply:
I'd like to see Ruby switch to something closer to what we're proposing here. Dumping all wrapper information into the platform field is a bit of a mess.

Reply:
I made a "Deviations" section that accommodates this, while still proposing an opinionated default for drivers who haven't done this yet.

Reply:
> I'd like to see Ruby switch to something closer to what we're proposing here. Dumping all wrapper information into the platform field is a bit of a mess.

@behackett: I think there's a misunderstanding here. I was only referring to Ruby possibly having an API to append data for a higher level library (e.g. Mongoid), which may be something to keep instead of moving over to DriverInfoOptions.

I was _not_ suggesting that Ruby continue to append Mongoid details in the platform field.